R/preprocessing.R

```diff
@@ -1526,3 +1526,165 @@ RegressOutNB <- function(
+    } else {
+        created_cell_col = FALSE
+    }
+
+    ## Check that umi.column exists in meta.data
+    if (!(umi.column %in% colnames(object@meta.data))) {
+        stop("`umi.column` specified not present in Seurat object provided")
+    }
+
+    ## Grab meta data
+    tmp_meta <- object@meta.data %>%
+        as.data.frame() %>%
+        dplyr::group_by(orig.ident) %>%
+        dplyr::mutate(rank = rank(-!!rlang::sym(umi.column))) %>%
+        dplyr::arrange(orig.ident, rank)
+
+    ## Set thresholds for rank of barcodes to ignore
+    if (is.null(threshold.low)) {
+        threshold.low <- 1
+    }
+    if (is.null(threshold.high)) {
+        threshold.high <- max(tmp_meta$rank)
+    }
+
+    ## Calculate inflection point within each group's barcode range
+    inflections <- tmp_meta %>%
+        tidyr::nest() %>%
+        dplyr::mutate(inflection = purrr::map_int(data, function(x) {
+            y <- x %>%
+                dplyr::mutate(
+                    rawdiff = c(NA, diff(log10(!!rlang::sym(umi.column) + 1)) / diff(rank))
+                ) %>%
+                dplyr::summarise(
+                    inflection = which(rawdiff == min(rawdiff[threshold.low:threshold.high], na.rm = TRUE))
+                )
+            y[[1]]
+        })) %>%
+        dplyr::select(-data)
+
+    ## Calculate inflection point plot for QC
+    if (show.plot == TRUE | return.plot == TRUE) {
+        g <- tmp_meta %>%
+            ggplot2::ggplot(aes(x = rank, y = log10(!!rlang::sym(umi.column)),
+                       group = orig.ident, colour = orig.ident)) +
+            ggplot2::geom_line() +
+            ggplot2::theme_bw() +
+            ggplot2::geom_vline(data = inflections,
+                               aes(xintercept = inflection,
+                                   group = orig.ident, colour = orig.ident),
+                               linetype = 'dashed')
+    }
+
+    ## Show the plot if desired
+    if (show.plot == TRUE) {
+        print(g)
+    }
+
+    ## Break: return ggplot2 object
+    if (return.plot == TRUE) {
+        return(g)
+    }
+
+    ## Filter the cells based on calculated inflection points per sample
+    cells_filtered <- tmp_meta %>%
+        dplyr::select(orig.ident, cell, !!rlang::sym(umi.column), rank) %>%
+        dplyr::group_by(orig.ident) %>%
+        tidyr::nest() %>%
+        dplyr::inner_join(inflections, by = 'orig.ident') %>%
+        dplyr::mutate(
+            umi_threshold = purrr::map2_dbl(data, inflection, function(d, i) {
+                x <- dplyr::filter(d, rank < i)
+                min(x[umi.column], na.rm = TRUE) # return the min nCount threshold used
+            }),
+            cells_pass = purrr::map2(data, inflection, function(d, i) {
+                x <- dplyr::filter(d, rank < i)
+                return(x$cell) # return passing cells
+            }))
+
+    ## Break: returns the calculated inflection points data and passing cells
+    if (return.inflections) {
```

Comment:
Functions should strive for a single return value that isn't dependent on a parameter, could you put all extra things in object meta data or tools?

Reply:
For this case, since it is returning the inflection point cut-off per sample, would it be more appropriate then to return it to tools (in meta data, it would be pretty repetitious to have it be appended on a per cell basis).

Reply:
Do whichever you think is best. The tools slot is designed to allow developers an unstructured free space to put whatever they want. We currently don't have setters/accessors for the tools slot, so use the `slot` function to access it. We'll take care of creating setters/accessors and updating code in Seurat to use such helper functions once we build them.